### PR TITLE
Bugfix/loc catalog performance

### PR DESCRIPTION
**Pull request by MikeNeilson**
Still needs work. Performs reasonably fast now, however the results aren't sorting correctly and I'm pretty sure the alias search won't work correctly. Plus there needs to be a minor difference between the with clause condition and the final condition.

### CODE REVIEW

**Comment by MikeNeilson** [comment]
Okay, got it to ~22 seconds on DMZ system and about ~4 on CWBI.

Additional i'm pretty sure the loc group stuff won't work right but I'm going to futz with that later, for now we just need the api itself working more reasonably fast.

**Comment by MikeNeilson** [comment]
Okay, got it to ~22 seconds on DMZ system and about ~4 on CWBI.

Additional i'm pretty sure the loc group stuff won't work right but I'm going to futz with that later, for now we just need the api itself working more reasonably fast.

File: cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java [hunk 1]
```diff
@@ -257,43 +265,91 @@ public Catalog getLocationCatalog(String cursor, int pageSize, String unitSystem
...
         //location codes (previous implementation used location_id) for joins, feel free to implement.
+        final AV_LOC2 avLoc2 = AV_LOC2.AV_LOC2;
         String locCursor = "*";
-        Condition condition = buildCatalogWhere(unitSystem, office, idLike, categoryLike, groupLike);
+        String searchOffice = office;
+        String curOffice = null;
+        Catalog.CatalogPage catPage = null;
         int total;
-        if (cursor == null || cursor.isEmpty()) {
+        Condition condition = avLoc2.LOCATION_ID.upper().likeRegex(idLike.toUpperCase())
```
Comment (adamkorynta) [inline cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java:278]: `idLike` needs a null check given the overload (or remove the overload if unused?)

File: cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java [hunk 2]
```diff
@@ -257,43 +265,91 @@ public Catalog getLocationCatalog(String cursor, int pageSize, String unitSystem
...
+                              .and(avLoc2.LOCATION_CODE.notEqual(0L))
```
Comment (adamkorynta) [inline cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java]: What is the 0 location code? Might be worth a comment explanation

File: cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java [hunk 3]
```diff
@@ -257,43 +265,91 @@ public Catalog getLocationCatalog(String cursor, int pageSize, String unitSystem
...
+            if( categoryLike == null && groupLike == null) {
+                condition = condition.and(avLoc2.ALIASED_ITEM.isNull());
+            }
+            if (searchOffice != null) {
+                condition =
+                        condition.and(avLoc2.DB_OFFICE_ID.upper().eq(searchOffice.toUpperCase()));
+            }
 
+            if (categoryLike != null) {
+                condition.and(avLoc2.LOC_ALIAS_CATEGORY.upper().likeRegex(categoryLike.toUpperCase()));
```
Comment (adamkorynta) [inline cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java]: reassign `condition`

File: cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java [hunk 4]
```diff
@@ -257,43 +265,91 @@ public Catalog getLocationCatalog(String cursor, int pageSize, String unitSystem
...
+            }
+
+            if (groupLike != null) {
+                condition.and(avLoc2.LOC_ALIAS_GROUP.upper().likeRegex(groupLike.toUpperCase()));
```
Comment (adamkorynta) [inline cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java]: reassign `condition`

File: cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java [hunk 5]
```diff
@@ -257,43 +265,91 @@ public Catalog getLocationCatalog(String cursor, int pageSize, String unitSystem
...
+            }
             SelectConditionStep<Record1<Integer>> count = dsl.select(count(asterisk()))
-                .from(AV_LOC2.AV_LOC2)
+                .from(avLoc2)
                 .where(condition);
-
+            logger.info(count.getSQL(ParamType.INLINED));
```
Comment (adamkorynta) [inline cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java]: looks like debug output, lower to finer?

File: cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java [hunk 6]
```diff
@@ -257,43 +265,91 @@ public Catalog getLocationCatalog(String cursor, int pageSize, String unitSystem
...
             total = count.fetchOne().value1();
         } else {
-            // get totally from page
-            Catalog.CatalogPage page = new Catalog.CatalogPage(cursor);
-            locCursor = page.getCursorId();
-            total = page.getTotal();
+            // get total from page
+            catPage = new Catalog.CatalogPage(page);
```
Comment (adamkorynta) [inline cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java:310]: I think this constructor needs a unit test. I'd argue it is in scope of this PR, but would be content with an Issue getting created.

File: cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java [hunk 7]
```diff
@@ -257,43 +265,91 @@ public Catalog getLocationCatalog(String cursor, int pageSize, String unitSystem
...
+
+        SelectSeekStep3<Record, String, ?, String> query = dsl.with(limiter).select(
+                limitId,
+                avLoc2.LOCATION_ID.as("alias_id"),
+                avLoc2.asterisk())
+            .from(limiter)
+            .leftOuterJoin(avLoc2).on(avLoc2.LOCATION_CODE.eq(limitCode))
+            .where(condition)            
+            .orderBy(avLoc2.DB_OFFICE_ID.asc(),limitId.asc(),avLoc2.ALIASED_ITEM.asc());
+        logger.info(query.getSQL(ParamType.INLINED));
```
Comment (adamkorynta) [inline cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java]: more debug output

File: cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java (hunk 2 above)
Comment (MikeNeilson) [reply cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java]: oh, the "DEleted TS", good point.

File: cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java [hunk 8]
```diff
@@ -257,43 +265,91 @@ public Catalog getLocationCatalog(String cursor, int pageSize, String unitSystem
...
-            .and(AV_LOC2.AV_LOC2.LOCATION_CODE.in(forLimit));
-
+        
+        if (curOffice != null) {
+            Condition officeEqualCur = avLoc2.DB_OFFICE_ID.upper().eq(curOffice.toUpperCase());
+            Condition curOfficeLocationIdGreater = avLoc2.LOCATION_ID.upper().gt(locCursor);
+            Condition officeGreaterThanCur = avLoc2.DB_OFFICE_ID.upper().gt(curOffice.toUpperCase());
+            condition = condition.and(officeEqualCur).and(curOfficeLocationIdGreater).or(officeGreaterThanCur);
+        } else {
+            condition.and(avLoc2.LOCATION_ID.upper().gt(locCursor));
```
Comment (adamkorynta) [inline cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java]: reassign `condition`

File: cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java (hunk 1 above)
Comment (MikeNeilson) [reply cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java:278]: Actually that needs a `@Notnull` or something.

File: cwms_radar_api/src/test/java/fixtures/RadarApiSetupCallback.java [hunk 9]
```diff
@@ -37,6 +37,9 @@ public class RadarApiSetupCallback implements BeforeAllCallback,AfterAllCallback
     private static CwmsDatabaseContainer cwmsDb;
     private static TestRealm realm;
     private static TestAuthValve authValve;
+    private static final String ORACLE_IMAGE = System.getProperty("RADAR.oracle.database.image", CwmsDatabaseContainer.ORACLE_18XE);
```
Comment (adamkorynta) [inline cwms_radar_api/src/test/java/fixtures/RadarApiSetupCallback.java:40]: I see new constants, but they aren't being used

File: cwms_radar_api/src/docker/Dockerfile [hunk 10]
```diff
@@ -15,6 +15,4 @@ ENV RADAR_POOL_MAX_ACTIVE "10"
 ENV RADAR_POOL_MAX_IDLE "5"
 ENV RADAR_POOL_MIN_IDLE "2"
 
-#EXPOSE 7000
-
-#CMD ["java","-jar","/radar/radar-standalone.jar"]
\ No newline at end of file
+EXPOSE 7000
```
Comment (adamkorynta) [inline cwms_radar_api/src/docker/Dockerfile:18]: Is this for debugging?

File: cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java (hunk 5 above)
Comment (MikeNeilson) [reply cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java]: yep.

File: cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java (hunk 6 above)
Comment (MikeNeilson) [reply cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java:310]: Good point, I'll see if I can do so quickly.

File: cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java (hunk 5 above)
Comment (adamkorynta) [reply cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java]: I've noticed a trend for info log levels, is it a pain to change those in a logging.properties file?

File: cwms_radar_api/src/docker/Dockerfile (hunk 10 above)
Comment (MikeNeilson) [reply cwms_radar_api/src/docker/Dockerfile:18]: no that tells the container user hey "this is the port you need to deal with."

File: cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java (hunk 5 above)
Comment (MikeNeilson) [reply cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java]: no, I'm just very lazy... and forgetful apparently.

File: cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java (hunk 1 above)
Comment (adamkorynta) [reply cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java:278]: that overload is still a problem though...

Comment (MikeNeilson) [reply cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java:278]: oy, now I see it.

File: cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java (hunk 8 above)
Comment (adamkorynta) [reply cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java]: just this last lonely condition

Comment (MikeNeilson) [reply cwms_radar_api/src/main/java/cwms/radar/data/dao/LocationsDaoImpl.java]: D'oh.

**Review by adamkorynta — APPROVED** [review]

**Pull request merged**